### ISSUE TITLE
Pylance does not honor union return type for __new__

### ISSUE THREAD

**Issue by manubb** [microsoft/pyright#6423]
## Environment data

-   Language Server version: 2023.11.10
-   OS and version: Ubuntu 23.04
-   Python version: 3.11.4

## Code Snippet

```python
from random import choice

class TriangleOrSquare:
    def __new__(cls, *args, **kwargs) -> Triangle | Square:
        # the right subclass cannot be resolved here in real user case,
        # it can only be after calling `super().__init__`
        return cast(Triangle | Square, super().__new__(cls))

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if choice([True, False]):  # Faking a condition
            self.__class__ = Triangle
        else:
            self.__class__ = Square


class Triangle(TriangleOrSquare):
    pass


class Square(TriangleOrSquare):
    pass


shape = TriangleOrSquare()  # (variable) shape: TriangleOrSquare
```

## Expected behavior

I'm expecting `shape` to have type `Triangle | Square`

## Actual behavior

`shape` has type `TriangleOrSquare`

## Other informations

When casting to a single subclass, it works as expected:
```python
class TriangleOrSquare:
    def __new__(cls, *args, **kwargs) -> Triangle:
        return cast(Triangle, super().__new__(cls))

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.__class__ = Triangle


class Triangle(TriangleOrSquare):
    pass

shape = TriangleOrSquare()  #(variable) shape: Triangle
```

**Comment by erictraut** [comment]
The behavior you're seeing is because of a bug in your code sample. The return type annotation for `__new__` references two classes (`Triangle` and `Square`) that are not yet defined at that point in the code. That means they are interpreted as `Unknown`, and pyright (the type checker upon which pylance is based) then assumes the standard behavior for `__new__`, which is to return an instance of its containing class.

If you run your code sample above, you'll see that you receive a runtime exception. Pylance will tell you about this error if you enable type checking diagnostics (i.e. set `python.analysis.typeCheckingMode` to "basic").

You can fix the bug in your code one of two ways:
1. Enclose the return type annotation in quotes. This is the mechanism prescribed in PEP 484 for forward-declared classes in type annotations.
2. Add a `from __future__ import annotations` to the top of your file. This mechanism "defers" evaluation of type annotations, allowing forward references to outer-scoped symbols to be resolved without a runtime exception.

**Comment by manubb** [comment]
Sorry. I did not include all the imports in the snippets. I have updated them.

**Comment by manubb** [comment]
@erictraut I have fixed the snippets. Can you consider reopening the issue please?

**Comment by erictraut** [comment]
Ah, I see what's going on. The behavior you're seeing was fixed in a recent version of pyright. The latest production version of pylance doesn't yet have the fix. If you switch to the prerelease version of pylance, you should see the behavior you're expecting.

**Comment by manubb** [comment]
Indeed, switching to `v2023.11.12` fixes the issue. Thanks for super fast support!
